WhateverGreen/kern_igfx_clock.cpp

```diff
@@ -0,0 +1,879 @@
+ *
+ *  @note This threshold is derived from the ICL framebuffer driver on macOS 10.15.6.
+ */
+static constexpr uint32_t ICL_CDCLK_DEC_FREQ_THRESHOLD = ICL_CDCLK_FREQ_648_0;
+
+/**
+ *  Core Display Clock PLL frequency in Hz for the 24 MHz hardware reference frequency
+ *
+ *  Main Reference:
+ *  - Intel Graphics Developer Manaual for Ice Lake Platforms, Volume 12 Display Engine
+ *    Page 171, CDCLK PLL Ratio and Divider Programming and Resulting Frequencies.
+ *
+ *  Side Reference:
+ *  - Intel Graphics Driver for Linux (5.8.3) bxt_calc_cdclk_pll_vco() in intel_cdclk.c
+ *
+ *  @note 54 is the PLL ratio when the reference frequency is 24 MHz
+ */
+static constexpr uint32_t ICL_CDCLK_PLL_FREQ_REF_24_0 = 24000000 * 54;
+
+/**
+ *  Core Display Clock PLL frequency in Hz for the 19.2 MHz hardware reference frequency
+ *
+ *  @note 68 is the PLL ratio when the reference frequency is 19.2 MHz
+ */
+static constexpr uint32_t ICL_CDCLK_PLL_FREQ_REF_19_2 = 19200000 * 68;
+
+/**
+ *  Core Display Clock PLL frequency in Hz for the 38.4 MHz hardware reference frequency
+ *
+ *  @note 34 is the PLL ratio when the reference frequency is 19.2 MHz
+ */
+static constexpr uint32_t ICL_CDCLK_PLL_FREQ_REF_38_4 = 38400000 * 34;
+
+// MARK: Patch Submodule IMP
+
+void IGFX::CoreDisplayClockFix::init() {
+	// We only need to patch the framebuffer driver
+	requiresPatchingGraphics = false;
+	requiresPatchingFramebuffer = true;
+}
+
+void IGFX::CoreDisplayClockFix::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	// Enable the Core Display Clock patch on ICL platforms
+	enabled = checkKernelArgument("-igfxcdc");
+	// Or if `enable-cdclk-frequency-fix` is set in IGPU property
+	if (!enabled)
+		enabled = info->videoBuiltin->getProperty("enable-cdclk-frequency-fix") != nullptr;
+}
+
+void IGFX::CoreDisplayClockFix::processFramebufferKext(KernelPatcher &patcher, size_t index, mach_vm_address_t address, size_t size) {
+	KernelPatcher::RouteRequest request = {
+		"__ZN31AppleIntelFramebufferController21probeCDClockFrequencyEv",
+		wrapProbeCDClockFrequency,
+		reinterpret_cast<mach_vm_address_t&>(orgProbeCDClockFrequency)
+	};
+	
+	orgDisableCDClock = patcher.solveSymbol<decltype(orgDisableCDClock)>(index, "__ZN31AppleIntelFramebufferController14disableCDClockEv", address, size);
+	orgSetCDClockFrequency = patcher.solveSymbol<decltype(orgSetCDClockFrequency)>(index, "__ZN31AppleIntelFramebufferController19setCDClockFrequencyEy", address, size);
+	orgIclReadRegister32 = reinterpret_cast<decltype(orgIclReadRegister32)>(callbackIGFX->AppleIntelFramebufferController__ReadRegister32);
+	
+	if (patcher.routeMultiple(index, &request, 1, address, size) && orgDisableCDClock && orgSetCDClockFrequency && orgIclReadRegister32) {
+		DBGLOG("igfx", "CDC: Functions have been routed successfully.");
+	} else {
+		patcher.clearError();
+		SYSLOG("igfx", "CDC: Failed to route functions.");
+	}
+}
+
```

Comment:
@vit9696 This is a follow-up to your review https://github.com/acidanthera/WhateverGreen/pull/72#discussion_r506909740.

I have updated my code to use `routeMultiple()` whenever possible, and now those functions are more readable.

However, I don't think the patcher supports solving multiple symbols in one shot. Take the above function as an example, the fix needs to resolve two more functions `disableCDClock()` and `setCDClockFreq()`. 

I would like to propose a new Lilu API `solveMultiple()` that is similar to `routeMultiple()`. The new API takes an array of `SolveRequest` and resolve all symbols in one function call. As a result, it would greatly simplify the case where multiple functions are required as helpers to perform a fix. 

What's your opinion?

Reply:
Hmmm, I do not mind, but I may not be able to work on it for the next week at least. If you send the patches we can make them land though.

Reply:
OK, I will do that and submit a PR to Lilu later.

Reply:
I have submitted a PR (https://github.com/acidanthera/Lilu/pull/61) to Lilu, and I also updated some code to use the new API in WhateverGreen. Finally, everything is clean and no more annoying type castings.